File: pm_core/src/infrastructure/store/mapper/workflow_transition_mapper.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package mapper

import (
	"github.com/serp/pm-core/src/core/domain/entity"
	"github.com/serp/pm-core/src/infrastructure/store/model"
)

type WorkflowTransitionMapper struct{}

func NewWorkflowTransitionMapper() *WorkflowTransitionMapper {
	return &WorkflowTransitionMapper{}
}

func (m *WorkflowTransitionMapper) ToEntity(mdl *model.WorkflowTransitionModel) *entity.WorkflowTransitionEntity {
	if mdl == nil {
		return nil
	}

	return &entity.WorkflowTransitionEntity{
		BaseEntity: entity.BaseEntity{
			ID:        mdl.ID,
			CreatedAt: mdl.CreatedAt.UnixMilli(),
			UpdatedAt: mdl.UpdatedAt.UnixMilli(),
		},
		WorkflowID:   mdl.WorkflowID,
		FromStateID:  mdl.FromStateID,
		ToStateID:    mdl.ToStateID,
		Name:         mdl.Name,
		ActiveStatus: mdl.ActiveStatus,
	}
}

func (m *WorkflowTransitionMapper) ToModel(e *entity.WorkflowTransitionEntity) *model.WorkflowTransitionModel {
	if e == nil {
		return nil
	}

	return &model.WorkflowTransitionModel{
		BaseModel: model.BaseModel{
			ID: e.ID,
		},
		WorkflowID:   e.WorkflowID,
		FromStateID:  e.FromStateID,
		ToStateID:    e.ToStateID,
		Name:         e.Name,
		ActiveStatus: e.ActiveStatus,
	}
}

func (m *WorkflowTransitionMapper) ToEntities(models []*model.WorkflowTransitionModel) []*entity.WorkflowTransitionEntity {
	if models == nil {
		return nil
	}

	entities := make([]*entity.WorkflowTransitionEntity, 0, len(models))
	for _, mdl := range models {
		if e := m.ToEntity(mdl); e != nil {
			entities = append(entities, e)
		}
	}

	return entities
}

File: pm_core/src/main/java/serp/project/pmcore/infrastructure/store/mapper/WorkflowTransitionMapper.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.pmcore.infrastructure.store.mapper;

import org.springframework.stereotype.Component;

import serp.project.pmcore.domain.workflow.entity.WorkflowTransitionEntity;
import serp.project.pmcore.infrastructure.store.model.WorkflowTransitionModel;

import java.util.Collections;
import java.util.List;

@Component
public class WorkflowTransitionMapper extends BaseMapper {

    public WorkflowTransitionModel toModel(WorkflowTransitionEntity entity) {
        if (entity == null) {
            return null;
        }
        return WorkflowTransitionModel.builder()
                .id(entity.getId())
                .tenantId(entity.getTenantId())
                .workflowVersionId(entity.getWorkflowVersionId())
                .name(entity.getName())
                .fromStepId(entity.getFromStepId())
                .toStepId(entity.getToStepId())
                .screenId(entity.getScreenId())
                .sequence(entity.getSequence())
                .createdAt(longToLocalDateTime(entity.getCreatedAt()))
                .createdBy(entity.getCreatedBy())
                .updatedAt(longToLocalDateTime(entity.getUpdatedAt()))
                .updatedBy(entity.getUpdatedBy())
                .deletedAt(longToLocalDateTime(entity.getDeletedAt()))
                .build();
    }

    public WorkflowTransitionEntity toEntity(WorkflowTransitionModel model) {
        if (model == null) {
            return null;
        }
        return WorkflowTransitionEntity.builder()
                .id(model.getId())
                .tenantId(model.getTenantId())
                .workflowVersionId(model.getWorkflowVersionId())
                .name(model.getName())
                .fromStepId(model.getFromStepId())
                .toStepId(model.getToStepId())
                .screenId(model.getScreenId())
                .sequence(model.getSequence())
                .createdAt(localDateTimeToLong(model.getCreatedAt()))
                .createdBy(model.getCreatedBy())
                .updatedAt(localDateTimeToLong(model.getUpdatedAt()))
                .updatedBy(model.getUpdatedBy())
                .deletedAt(localDateTimeToLong(model.getDeletedAt()))
                .build();
    }

    public List<WorkflowTransitionEntity> toEntities(List<WorkflowTransitionModel> models) {
        if (models == null || models.isEmpty()) {
            return Collections.emptyList();
        }
        return models.stream().map(this::toEntity).toList();
    }

    public List<WorkflowTransitionModel> toModels(List<WorkflowTransitionEntity> entities) {
        if (entities == null || entities.isEmpty()) {
            return Collections.emptyList();
        }
        return entities.stream().map(this::toModel).toList();
    }
}
